libacpi.3
.hy 0
.TH libacpi 3 LIBACPI ""
.ds n 5
.ds d /etc/terminfo
.SH NAME
\fBlibacpi\fR \- general purpose library for ACPI
.SH SYNOPSIS
\fB#include <libacpi.h>\fR
.br
.SH DESCRIPTION
The \fBlibacpi\fR library routines provide its users a method
of reading ACPI values in a comfortable way. This includes
battery information, fan states, thermal zones and AC states.
.PP
A program using these routines must be linked with \fB\-lacpi\fR option.
.PP
To initialize this functionality you have to first allocate space for a global_t structure
.sp
    \fBglobal_t *global = malloc (sizeof (global_t));\fR
.sp

Now use the \fBcheck_acpi_support()\fR function to check if the system actually supports ACPI.
In case it does the function will return \fBSUCCESS\fR or \fBNOT_SUPPORTED\fR on failure.
.sp
Now you can use the following functions to initialize the values you get
from the ACPI interface:
.sp
    \fBinit_acpi_batt(global);\fR
    \fBinit_acpi_acadapt(global);\fR
    \fBinit_acpi_fan(global);\fR
    \fBinit_acpi_thermal(global);\fR
.sp
To know if some of the features is not supported it is a good idea to store the return values
in some variable.
.sp
If you want to read non\-static values the ACPI interface provides (for example battery percentage), use the following functions to do this:
.sp
    \fBread_acpi_batt(int num)\fR
    \fBread_acpi_zone(int num, global_t *globals);\fR
    \fBread_acpi_fan(int num)\fR
    \fBread_acpi_acstate(global_t *globals);\fR
.sp
Since you can have more than one battery, fan and thermal zone, num is the number you want to read, for example in a loop.
.sp
    \fBthermal_t *tp;\fR
    ....
    \fBfor(i=0; i<global\->thermal_count; i++){\fR
    \fB    read_acpi_zone(i, global);\fR
    \fB    tp = &thermals[i];\fR
    ...
    \fB}\fR
.SS "Structures"

.in +1c
.ti -1c
.RI "struct \fBfan_t\fP"
.br
.RI "\fIfan data \fP"
.ti -1c
.RI "struct \fBbattery_t\fP"
.br
.RI "\fIinformation found about battery \fP"
.ti -1c
.RI "struct \fBthermal_t\fP"
.br
.RI "\fIinformation about thermal zone \fP"
.ti -1c
.RI "struct \fBadapter_t\fP"
.br
.RI "\fIinformation about ac adapater \fP"
.ti -1c
.RI "struct \fBglobal_t\fP"
.br
.RI "\fIglobal acpi structure \fP"
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "int \fBinit_acpi_batt\fP (\fBglobal_t\fP *globals)"
.br
.ti -1c
.RI "int \fBinit_acpi_acadapt\fP (\fBglobal_t\fP *globals)"
.br
.ti -1c
.RI "int \fBinit_acpi_thermal\fP (\fBglobal_t\fP *globals)"
.br
.ti -1c
.RI "int \fBinit_acpi_fan\fP (\fBglobal_t\fP *globals)"
.br
.ti -1c
.RI "int \fBcheck_acpi_support\fP (void)"
.br
.ti -1c
.RI "int \fBread_acpi_batt\fP (const int num)"
.br
.ti -1c
.RI "void \fBread_acpi_acstate\fP (\fBglobal_t\fP *globals)"
.br
.ti -1c
.RI "int \fBread_acpi_zone\fP (const int num, \fBglobal_t\fP *globals)"
.br
.ti -1c
.RI "int \fBread_acpi_fan\fP (const int num)"
.br
.in -1c
.SS "Variables"

.in +1c
.ti -1c
.RI "\fBbattery_t\fP \fBbatteries\fP [MAX_ITEMS]"
.br
.ti -1c
.RI "\fBthermal_t\fP \fBthermals\fP [MAX_ITEMS]"
.br
.ti -1c
.RI "\fBfan_t\fP \fBfans\fP [MAX_ITEMS]"
.br
.in -1c
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum { \fBITEM_EXCEED\fP =  \-5, \fBDISABLED\fP =  \-4, \fBNOT_PRESENT\fP =  \-3, \fBALLOC_ERR\fP =  \-2, \fBNOT_SUPPORTED\fP =  \-1, \fBSUCCESS\fP }"
.br
.ti -1c
.RI "enum \fBpower_state_t\fP { \fBP_AC\fP, \fBP_BATT\fP, \fBP_ERR\fP }"
.br
.ti -1c
.RI "enum \fBthermal_state_t\fP { \fBT_CRIT\fP, \fBT_HOT\fP, \fBT_PASS\fP, \fBT_ACT\fP, \fBT_OK\fP, \fBT_ERR\fP }"
.br
.ti -1c
.RI "enum \fBcharge_state_t\fP { \fBC_CHARGE\fP, \fBC_DISCHARGE\fP, \fBC_CHARGED\fP, \fBC_NOINFO\fP, \fBC_ERR\fP }"
.br
.ti -1c
.RI "enum \fBbatt_state_t\fP { \fBB_HIGH\fP, \fBB_MED\fP, \fBB_LOW\fP, \fBB_CRIT\fP, \fBB_HARD_CRIT\fP, \fBB_ERR\fP }"
.br
.ti -1c
.RI "enum \fBthermal_mode_t\fP { \fBCO_ACT\fP, \fBCO_PASS\fP, \fBCO_CRIT\fP, \fBCO_ERR\fP }"
.br
.ti -1c
.RI "enum \fBfan_state_t\fP { \fBF_ON\fP, \fBF_OFF\fP, \fBF_ERR\fP }"
.br
.in -1c
.SH "Detailed Description"
.PP
Now follows a more detailed description of functions, structures and enums.

Definition in file \fBlibacpi.h\fP.
.SH "Function Documentation"
.PP
.SS "int check_acpi_support (void)"
.PP
Checks if the system does support ACPI or not.
.PP
\fBReturns:\fP
.RS 4
SUCCESS if the system supports ACPI or, NOT_SUPPORTED.
.RE
.SS "int init_acpi_acadapt (\fBglobal_t\fP * globals)"
.PP
Finds existing ac adapter and fills the adapt (in globa struct) structure with the paths of the important files to parse.
.PP
\fBParameters:\fP
.RS 4
\fIglobals\fP pointer to global acpi structure
.PP
\fBReturns:\fP
.Rs 4
SUCCESS if everything is ok, NOT_SUPPORTED if no information can be found, ALLOC_ERR on allocation errors.
.RE
.PP
.SS "int init_acpi_batt (\fBglobal_t\fP * globals)"
.PP
Finds existing batteries and fills the corresponding batteries structures with the paths of the important files to parse.
.PP
\fBParameters:\fP
.RS 4
\fIglobals\fP pointer to global acpi structure
.PP
\fBReturns:\fP
.Rs 4
SUCCESS if everything is ok, NOT_SUPPORTED if no information can be found, ALLOC_ERR on allocation errors and ITEM_EXCEED if
there are more batteries found than MAX_ITEMS, this should never happen.
.RE
.PP
.SS "int init_acpi_fan (\fBglobal_t\fP * globals)"
.PP
Finds existing fans and fills corresponding fan structures with the paths of the important files to parse for fan information.
.PP
\fBParameters:\fP
.RS 4
\fIglobals\fP pointer to global acpi structure
.PP
\fBReturns:\fP
.Rs 4
SUCCESS if everything is ok, NOT_SUPPORTED if no information can be found, ALLOC_ERR on allocation errors and ITEM_EXCEED if
there are more fans found than MAX_ITEMS, this should never happen.
.RE
.PP
.SS "int init_acpi_thermal (\fBglobal_t\fP * globals)"
.PP
Finds existing thermal zones and fills corresponding thermal structures with the paths of the important files to parse for thermal information.
.PP
\fBParameters:\fP
.RS 4
\fIglobals\fP pointer to global acpi structure
.PP
\fBReturns:\fP
.Rs 4
SUCCESS if everything is ok, NOT_SUPPORTED if no information can be found, ALLOC_ERR on allocation errors and ITEM_EXCEED if
there are more thermal zones found than MAX_ITEMS, this should never happen.
.RE
.PP
.SS "void read_acpi_acstate (\fBglobal_t\fP * globals)"
.PP
Looks up if the ac adapter is plugged in or not and sets the values in a struct
.PP
\fBParameters:\fP
.RS 4
\fIglobals\fP pointer to the global acpi structure
.RE
.PP
Referenced by init_acpi_acadapt().
.SS "int read_acpi_batt (const int num)"
.PP
Gathers all information of a given battery and filling a struct with it.
.PP
\fBParameters:\fP
.RS 4
\fInum\fP number of battery
.PP
\fBReturns:\fP
.Rs 4
SUCCESS if everything is ok or ITEM_EXCEED if there is not battery num.
.SS "int read_acpi_fan (const int num)"
.PP
Gathers all information about given fan and sets the corresponding values in a struct.
.PP
\fBParameters:\fP
.RS 4
\fInum\fP number for the fan to read
.PP
\fBReturns:\fP
.Rs 4
SUCCESS if everything is ok, ITEM_EXCEED if there is not fan num or NOT_SUPPORTED if the
values can't be read. This should not happen if the init function returned SUCCESS if the ACPI implementation.
.SS "int read_acpi_zone (const int num, \fBglobal_t\fP * globals)"
.PP
Gathers all information of a given thermal zone and sets the corresponding values in a struct.
.PP
\fBParameters:\fP
.RS 4
\fInum\fP zone
.br
\fIglobals\fP pointer to global acpi struct, needed if there is just one zone
.PP
\fBReturns:\fP
.Rs 4
SUCCESS if everything is ok, ITEM_EXCEED if there is not thermal zone num or NOT_SUPPORTED if the
values can't be read. This should not happen if the init function returned SUCCESS if the ACPI implementation.

.SH "Enumeration Type Documentation"
.PP
.SS "possible function return values"
.PP
\fBEnumerator: \fP
.in +1c
.TP
\fB\fIITEM_EXCEED \fP\fP
maximum item count reached
.TP
\fB\fIDISABLED \fP\fP
features is disabled
.TP
\fB\fINOT_PRESENT \fP\fP
something is not present
.TP
\fB\fIALLOC_ERR \fP\fP
an error occurred while allocating space
.TP
\fB\fINOT_SUPPORTED \fP\fP
a feature is not supported
.TP
\fB\fISUCCESS \fP\fP
everything was fine
.SS "enum \fBbatt_state_t\fP"
.PP
\fBEnumerator: \fP
.in +1c
.TP
\fB\fIB_HIGH \fP\fP
remaining battery life is high
.TP
\fB\fIB_MED \fP\fP
remaining battery life is medium
.TP
\fB\fIB_LOW \fP\fP
remaining battery life is low
.TP
\fB\fIB_CRIT \fP\fP

CHANGES
0.2 (2007-07-29):
    * Fixed memleaks
    * Prevent double header inclusion, thanks Julien Blache

0.1 (2007-06-14):
    * Initial release


libacpi.c
/*
 * (C)opyright 2007 Nico Golde <nico@ngolde.de>
 * See LICENSE file for license details
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <dirent.h>
#include <ctype.h>
#include <stddef.h>

#include "libacpi.h"
#include "list.h"


static int read_acpi_battinfo(const int num, const int sysstyle);
static int read_acpi_battalarm(const int num, const int sysstyle);
static int read_acpi_battstate(const int num);
static void read_acpi_thermalzones(global_t *globals);

typedef struct {
	char * value;
	size_t offset;
} acpi_value_t;

static acpi_value_t
battinfo_values[] = {
	{ "last full capacity:", offsetof(battery_t, last_full_cap) },
	{ "design voltage:", offsetof(battery_t, design_voltage) },
	{ "design capacity warning:", offsetof(battery_t, design_warn) },
	{ "design capacity low:", offsetof(battery_t, design_low) },
	{ "capacity granularity 1:", offsetof(battery_t, design_level1) },
	{ "capacity granularity 2:", offsetof(battery_t, design_level2) },
	{ NULL, 0 }
};

/* given a buffer for example from a file, search for key
 * and return a pointer to the value of it. On error return NULL*/
static char *
scan_acpi_value(const char *buf, const char *key){
	char *ptr = NULL;
	char *tmpbuf = NULL;
	char *tmpkey = NULL;
	char *tmpval = NULL;

	if((tmpbuf = strdup(buf)) == NULL)
		return NULL;

	/* jump to the key in buffer */
	if((tmpkey = strstr(tmpbuf, key))) {
		/* jump behind the key, whitespaces and tabs */
		for(tmpkey += strlen(key); *tmpkey && (*tmpkey == ' ' || *tmpkey == '\t'); tmpkey++);
		for(tmpval = tmpkey; *tmpval && *tmpval != ' ' &&
				*tmpval != '\t' && *tmpval != '\n' &&
				*tmpval != '\r'; tmpval++);
		if(tmpval)
			*tmpval = '\0';

		if((ptr = strdup(tmpkey)) == NULL) {
			free(tmpbuf);
			return NULL;
		}
	}
	free(tmpbuf);
	return ptr;
}

/* reads a file into a buffer and returns a pointer to it, or NULL on error */
static char *
get_acpi_content(const char *file){
	FILE *input = NULL;
	char *buf = NULL;
	int read = 0;

	if((buf = malloc(MAX_BUF + 1)) == NULL)
		return NULL;
	if((input = fopen(file, "r")) == NULL) {
        free(buf);
		return NULL;
    }
	read = fread(buf, 1, MAX_BUF, input);
	if(read > 0) buf[read - 1] = '\0';
	else buf[0] = '\0'; /* I would consider it a kernel bug if that happens */

	fclose(input);
	return buf;
}

/* returns the acpi version or NOT_SUPPORTED(negative value) on failure */
static int
get_acpi_version(void){
	long ret = -1;
	char *tmp = get_acpi_content(PROC_ACPI "info");
	char *version = NULL;
	
	if(!tmp) {
		tmp = get_acpi_content("/sys/module/acpi/parameters/acpica_version");
		if (tmp) {
			long ret = strtol(tmp, NULL, 10);
			free(tmp);
			return ret;
		} else {
			return NOT_SUPPORTED;
		}
	}
	if((version = scan_acpi_value(tmp, "version:")) == NULL){
		free(tmp);
		return NOT_SUPPORTED;
	}
	ret = strtol(version, NULL, 10);
	free(tmp);
	free(version);
	return ret;
}

/* check if acpi is supported on the system, return 0 on success
 * and -1 if not */
int
check_acpi_support(void){
	int version = get_acpi_version();

	/* we don't support 2.4 kernel versions TODO */
	if(version == NOT_SUPPORTED || version < 20020214)
		return NOT_SUPPORTED;
	return SUCCESS;
}

/* reads existent battery directories and starts to fill the battery
 * structure. Returns 0 on success, negative values on error */
int
init_acpi_batt(global_t *globals){
	char *names[MAX_ITEMS];
	battery_t *binfo;
	list_t *lst = NULL;
	node_t *node = NULL;
	int dir_count, num_bat, i = 0;

	globals->batt_count = 0;
	globals->sysstyle = 0;
	if((lst = dir_list(PROC_ACPI "battery")) == NULL || !lst->top)
	{
        if (lst)
            delete_list(lst);
		/* check for new Linux 2.6.24+ layout */
		if((lst = dir_list(SYS_POWER)) == NULL || !lst->top)
			return NOT_SUPPORTED;
		else
			globals->sysstyle = 1;
	}
	for(node = lst->top; node; node=node->next){
		if((names[globals->batt_count] = strdup(node->name)) == NULL){
			delete_list(lst);
			return ALLOC_ERR;
		}
        globals->batt_count++;
	}

	if(globals->batt_count > MAX_ITEMS) return ITEM_EXCEED;

	/* A quick insertion sort, to sort battery names */
	{
		char *tmp1, *tmp2;
		int x,y;
		for (x = 1; x < globals->batt_count; x++) {
			tmp1 = names[x];
			y = x - 1;
			while ((y >= 0) && ((strcmp (tmp1, names[y])) < 0)) {
				tmp2 = names[y + 1];
				names[y + 1] = names[y];
				names[y] = tmp2;
			}
		}
	}

    dir_count = globals->batt_count;

    for (num_bat=i=0; i < dir_count && i < MAX_ITEMS; i++){
        if (strncmp(names[i], "BAT", 3) == 0) {
            binfo = &batteries[num_bat];
            snprintf(binfo->name, MAX_NAME, "%s", names[i]);
            if(globals->sysstyle) {
                    snprintf(binfo->state_file, MAX_NAME, SYS_POWER "/%s/status", names[i]);
                    snprintf(binfo->info_file, MAX_NAME, SYS_POWER "/%s", names[i]);
                    snprintf(binfo->alarm_file, MAX_NAME, SYS_POWER "/%s/alarm", names[i]);
            } else {
                    snprintf(binfo->state_file, MAX_NAME, PROC_ACPI "battery/%s/state", names[i]);
                    snprintf(binfo->info_file, MAX_NAME, PROC_ACPI "battery/%s/info", names[i]);
                    snprintf(binfo->alarm_file, MAX_NAME, PROC_ACPI "battery/%s/alarm", names[i]);
            }
            read_acpi_battinfo(num_bat, globals->sysstyle);
            read_acpi_battalarm(num_bat, globals->sysstyle);
            num_bat++;
        } else {
            globals->batt_count--;
        }
		free(names[i]);
	}
	delete_list(lst);
	return SUCCESS;
}

/* reads the acpi state and writes it into the globals structure, void */
void
read_acpi_acstate(global_t *globals){
	adapter_t *ac = &globals->adapt;
	char *buf = NULL;
	char *tmp = NULL;

	if(ac->state_file && (buf = get_acpi_content(ac->state_file)) == NULL){
		ac->ac_state = P_ERR;
		return;
	}
	if(globals->sysstyle)
	{
		if(!strcmp(buf, "1"))
			ac->ac_state = P_AC;
		else if(!strcmp(buf, "0"))
			ac->ac_state = P_BATT;
		else ac->ac_state = P_ERR;
	}
	else
	{
		if((tmp = scan_acpi_value(buf, "state:")) && !strncmp(tmp, "on-line", 7))
			ac->ac_state = P_AC;
		else if(tmp && !strncmp(tmp, "off-line", 8))
			ac->ac_state = P_BATT;
		else ac->ac_state = P_ERR;
	}
	free(buf);
	free(tmp);
}

/* reads the name of the ac-adapter directory and fills the adapter_t
 * structure with the name and the state-file. Return 0 on success, negative values on errors */
int
init_acpi_acadapt(global_t *globals){
	list_t *lst = NULL;
	adapter_t *ac = &globals->adapt;

	globals->sysstyle = 0;
	if((lst = dir_list(PROC_ACPI "ac_adapter")) == NULL || !lst->top)
	{
		if((lst = dir_list(SYS_POWER "/AC")) == NULL || !lst->top)
			return NOT_SUPPORTED;
		else
			globals->sysstyle = 1;
	}
	if((!lst->top->name || ((ac->name = strdup(lst->top->name)) == NULL))){
		delete_list(lst);
		return ALLOC_ERR;
	}
	if(globals->sysstyle)
		snprintf(ac->state_file, MAX_NAME, SYS_POWER "/AC/online");
	else
		snprintf(ac->state_file, MAX_NAME, PROC_ACPI "ac_adapter/%s/state", ac->name);
	delete_list(lst);
	read_acpi_acstate(globals);
	return SUCCESS;
}

/* read acpi information for fan num, returns 0 on success and negative values on errors */
int
read_acpi_fan(const int num){
	char *buf = NULL;
	char *tmp = NULL;
	fan_t *info = &fans[num];

	if(num > MAX_ITEMS) return ITEM_EXCEED;

	/* scan state file */
	if((buf = get_acpi_content(info->state_file)) == NULL)
		info->fan_state = F_ERR;

	if(!buf || (tmp = scan_acpi_value(buf, "status:")) == NULL){
		info->fan_state = F_ERR;
		return NOT_SUPPORTED;
	}
	if (tmp[0] == 'o' && tmp[1] == 'n') info->fan_state = F_ON;
	else if(tmp[0] == 'o' && tmp[1] == 'f') info->fan_state = F_OFF;
	else info->fan_state = F_ERR;
	free(buf);
	free(tmp);
	return SUCCESS;
}

/* read all fans, fill the fan structures */
static void
read_acpi_fans(global_t *globals){
	int i;
	for(i = 0; i < globals->fan_count; i++)
		read_acpi_fan(i);
}

/* reads the names of the fan directories, fills fan_t,
 * return 0 on success, negative values on errors */
int
init_acpi_fan(global_t *globals){
	char *names[MAX_ITEMS];
	list_t *lst = NULL;
	node_t *node = NULL;
	int i = 0;
	fan_t *finfo = NULL;
	globals->fan_count = 0;

	if((lst = dir_list(PROC_ACPI "fan")) == NULL || !lst->top)
		return NOT_SUPPORTED;
	for(node = lst->top; node; node = node->next){
		if((names[globals->fan_count] = strdup(node->name)) == NULL){
			delete_list(lst);
			return ALLOC_ERR;
		}
		globals->fan_count++;
	}

	if(globals->fan_count > MAX_ITEMS) return ITEM_EXCEED;

	for (; i < globals->fan_count && i < MAX_ITEMS; i++){
		finfo = &fans[i];
		snprintf(finfo->name, MAX_NAME, "%s", names[i]);
		snprintf(finfo->state_file, MAX_NAME, PROC_ACPI "fan/%s/state", names[i]);
		free(names[i]);
	}
	delete_list(lst);
	read_acpi_fans(globals);
	return SUCCESS;
}

/* reads the name of the thermal-zone directory and fills the adapter_t
 * structure with the name and the state-file. Return 0 on success, negative values on errors */
int
init_acpi_thermal(global_t *globals){
	char *names[MAX_ITEMS];
	list_t *lst = NULL;
	node_t *node = NULL;
	thermal_t *tinfo = NULL;
	int i = 0;
	globals->thermal_count = 0;

	if((lst = dir_list(PROC_ACPI "thermal_zone")) == NULL)
		return NOT_SUPPORTED;
	for(node = lst->top; node; node = node->next){
		if((names[globals->thermal_count] = strdup(node->name)) == NULL){
			delete_list(lst);
			return ALLOC_ERR;
		}
		globals->thermal_count++;
	}

	if(globals->thermal_count > MAX_ITEMS) return ITEM_EXCEED;

	for (; i < globals->thermal_count && i < MAX_ITEMS; i++){
		tinfo = &thermals[i];
		snprintf(tinfo->name, MAX_NAME, "%s", names[i]);
		snprintf(tinfo->state_file, MAX_NAME, PROC_ACPI "thermal_zone/%s/state", names[i]);
		snprintf(tinfo->temp_file, MAX_NAME, PROC_ACPI "thermal_zone/%s/temperature", names[i]);
		snprintf(tinfo->cooling_file, MAX_NAME, PROC_ACPI "thermal_zone/%s/cooling_mode", names[i]);
		snprintf(tinfo->freq_file, MAX_NAME, PROC_ACPI "thermal_zone/%s/polling_frequency", names[i]);
		snprintf(tinfo->trips_file, MAX_NAME, PROC_ACPI "thermal_zone/%s/trip_points", names[i]);
		free(names[i]);
	}
	delete_list(lst);
	read_acpi_thermalzones(globals);
	return SUCCESS;
}

/* checks the string state and sets the thermal state, returns void */
static void
thermal_state(const char *state, thermal_t *info){
	if(state[0] == 'o')
		info->therm_state = T_OK;
	else if(!strncmp (state, "crit", 4))
		info->therm_state = T_CRIT;
	else if (!strncmp (state, "hot", 3))
		info->therm_state = T_HOT; else if (!strncmp (state, "pas", 3))
		info->therm_state = T_PASS;
	else
		info->therm_state = T_ACT;
}

/* checks the string tmp and sets the cooling mode */
static void
fill_cooling_mode(const char *tmp, thermal_t *info){
	if(tmp[0] == 'a')
		info->therm_mode = CO_ACT;
	else if(tmp[0]  == 'p')
		info->therm_mode = CO_PASS;
	else info->therm_mode = CO_CRIT;
}

/* reads values for thermal_zone num, return 0 on success, negative values on error */
int
read_acpi_zone(const int num, global_t *globals){
	char *buf = NULL;
	char *tmp = NULL;
	thermal_t *info = &thermals[num];

	if(num > MAX_ITEMS) return ITEM_EXCEED;

	/* scan state file */
	if((buf = get_acpi_content(info->state_file)) == NULL)
		info->therm_state = T_ERR;

	if(buf && (tmp = scan_acpi_value(buf, "state:")))
			thermal_state(tmp, info);
	free(tmp);
	free(buf);

	/* scan temperature file */
	if((buf = get_acpi_content(info->temp_file)) == NULL)
		info->temperature = NOT_SUPPORTED;

	if(buf && (tmp = scan_acpi_value(buf, "temperature:"))){
		info->temperature = strtol(tmp, NULL, 10);
		/* if we just have one big thermal zone, this will be the global temperature */
		if(globals->thermal_count == 1)
			globals->temperature = info->temperature;
	}
	free(tmp);
	free(buf);

	/* scan cooling mode file */
	if((buf = get_acpi_content(info->cooling_file)) == NULL)
		info->therm_mode = CO_ERR;
	if(buf && (tmp = scan_acpi_value(buf, "cooling mode:")))
		fill_cooling_mode(tmp, info);
	else info->therm_mode = CO_ERR;
	free(tmp);
	free(buf);

	/* scan polling_frequencies file */
	if((buf = get_acpi_content(info->freq_file)) == NULL)
		info->frequency = DISABLED;
	if(buf && (tmp = scan_acpi_value(buf, "polling frequency:")))
		info->frequency = strtol(tmp, NULL, 10);
	else info->frequency = DISABLED;
	free(tmp);
	free(buf);

	/* TODO: IMPLEMENT TRIP POINTS FILE */

	return SUCCESS;
}

/* read all thermal zones, fill the thermal structures */
static void
read_acpi_thermalzones(global_t *globals){
	int i;
	for(i = 0; i < globals->thermal_count; i++)
		read_acpi_zone(i, globals);
}

/* fill battery_state for given battery, return 0 on success or negative values on error */
static void
batt_charge_state(battery_t *info){
	int high = info->last_full_cap / 2;
	int med = high / 2;

	if(info->remaining_cap > high)
		info->batt_state = B_HIGH;
	else if(info->remaining_cap <= high && info->remaining_cap > med)
		info->batt_state = B_MED;
	else if(info->remaining_cap <= med && info->remaining_cap > info->design_warn)
		info->batt_state = B_LOW;
	else if(info->remaining_cap <= info->design_warn && info->remaining_cap > info->design_low)
		info->batt_state = B_CRIT;
	else info->batt_state = B_HARD_CRIT;
}

/* fill charge_state of a given battery num, return 0 on success or negative values on error */
static charge_state_t
fill_charge_state(const char *state, battery_t *info){
    info->charge_state = C_NOINFO;
	if(state[0] == 'u')
		info->charge_state = C_ERR;
	else if(!strncasecmp (state, "disch", 5))
		info->charge_state = C_DISCHARGE;
	else if (!strncasecmp (state, "full", 4))
		info->charge_state = C_CHARGED;
	else if (!strncasecmp (state, "chargi", 6))
		info->charge_state = C_CHARGE;
	else
		info->charge_state = C_NOINFO;
    return info->charge_state;
}

/* read alarm capacity, return 0 on success, negative values on error */
static int
read_acpi_battalarm(const int num, const int sysstyle){
	char *buf = NULL;
	char *tmp = NULL;
	battery_t *info = &batteries[num];

	if((buf = get_acpi_content(info->alarm_file)) == NULL)
		return NOT_SUPPORTED;

	if(sysstyle)
	{
		if(!strcmp(buf, "0"))
			info->alarm = 0;
		else if(!strcmp(buf, "1"))
			info->alarm = 1;
		else
			info->alarm = NOT_SUPPORTED;
	}
	else
	{
		if((tmp = scan_acpi_value(buf, "alarm:")) && tmp[0] != 'u')
			info->alarm = strtol(tmp, NULL, 10);
		else
			info->alarm = NOT_SUPPORTED;
	}
	free(buf);
	free(tmp);
	return SUCCESS;
}

/* reads static values for a battery (info file), returns SUCCESS */
static int
read_acpi_battinfo(const int num, const int sysstyle){
	char *buf = NULL;
	char *tmp = NULL;
	battery_t *info = &batteries[num];
	int i = 0;
	char sysfile[MAX_NAME];

	if(sysstyle)
	{
		snprintf(sysfile, MAX_NAME, "%s/present", info->info_file);
		if((buf = get_acpi_content(sysfile)) == NULL)
			return NOT_SUPPORTED;
		if(!strcmp(buf, "1")) {
			info->present = 1;
		} else {
			info->present = 0;
			return NOT_PRESENT;
		}
        free(buf);

		snprintf(sysfile, MAX_NAME, "%s/charge_full_design", info->info_file);
		if((buf = get_acpi_content(sysfile)) == NULL)
			return NOT_SUPPORTED;
		info->design_cap = strtol(buf, NULL, 10);
        free(buf);

		snprintf(sysfile, MAX_NAME, "%s/charge_full", info->info_file);
		if((buf = get_acpi_content(sysfile)) == NULL)
			return NOT_SUPPORTED;
		info->last_full_cap = strtol(buf, NULL, 10);
        free(buf);

		snprintf(sysfile, MAX_NAME, "%s/charge_now", info->info_file);
		if((buf = get_acpi_content(sysfile)) == NULL)
			return NOT_SUPPORTED;
		info->remaining_cap = strtol(buf, NULL, 10);
        free(buf);

		snprintf(sysfile, MAX_NAME, "%s/voltage_min_design", info->info_file);
		if((buf = get_acpi_content(sysfile)) == NULL)
			return NOT_SUPPORTED;
		info->design_voltage = strtol(buf, NULL, 10);
        free(buf);

		snprintf(sysfile, MAX_NAME, "%s/voltage_now", info->info_file);
		if((buf = get_acpi_content(sysfile)) == NULL)
			return NOT_SUPPORTED;
		info->present_voltage = strtol(buf, NULL, 10);
        free(buf);

		/* FIXME: is rate == current here? */
		snprintf(sysfile, MAX_NAME, "%s/current_now", info->info_file);
		if((buf = get_acpi_content(sysfile)) == NULL)
			return NOT_SUPPORTED;
		info->present_rate = strtol(buf, NULL, 10);
        free(buf);

		return SUCCESS;
	}

	if((buf = get_acpi_content(info->info_file)) == NULL)
		return NOT_SUPPORTED;

	/* you have to read the present value always since a battery can be taken away while
	 * refreshing the data */
	if((tmp = scan_acpi_value(buf, "present:")) && !strncmp(tmp, "yes", 3)) {
		free(tmp);
		info->present = 1;
	} else {
		info->present = 0;
		free(buf);
		return NOT_PRESENT;
	}

	if((tmp = scan_acpi_value(buf, "design capacity:")) && tmp[0] != 'u'){
		info->design_cap = strtol(tmp, NULL, 10);
		/* workaround ACPI's broken way of reporting no battery */
		if(info->design_cap == 655350) info->design_cap = NOT_SUPPORTED;
		free(tmp);
	}
	else info->design_cap = NOT_SUPPORTED;

	for (;battinfo_values[i].value; i++) {
		if ((tmp = scan_acpi_value(buf, battinfo_values[i].value)) && tmp[0] != 'u') {
			*((int *)(((char *)info) + battinfo_values[i].offset)) = strtol(tmp, NULL, 10);
			free(tmp);
		} else {
			*((int *)(((char *)info) + battinfo_values[i].offset)) = NOT_SUPPORTED;
		}
	}

	/* TODO remove debug */
	/* printf("%s\n", buf); */
	free(buf);

	return SUCCESS;
}

/* read information for battery num, return 0 on success or negative values on error */
static int
read_acpi_battstate(const int num){
	char *buf = NULL;
	char sysfile[MAX_NAME];
    charge_state_t cstate;
	battery_t *info = &batteries[num];

	if((buf = get_acpi_content(info->state_file)) == NULL) {
		info->present = 0;
		return NOT_PRESENT;
    } else {
		info->present = 1;
	}

	/* TODO REMOVE DEBUG */
	/* printf("%s\n\n", buf); */

    cstate = fill_charge_state(buf, info);
    free(buf);
    if ((cstate == C_NOINFO) || (cstate == C_ERR)) {
        return NOT_SUPPORTED;
    }

    snprintf(sysfile, MAX_NAME, "%s/charge_now", info->info_file);
    if ((buf = get_acpi_content(sysfile)) != NULL)
        info->remaining_cap = strtol(buf, NULL, 10);
    free(buf);

    snprintf(sysfile, MAX_NAME, "%s/voltage_now", info->info_file);
    if ((buf = get_acpi_content(sysfile)) != NULL)
        info->present_voltage = strtol(buf, NULL, 10);
    free(buf);

    snprintf(sysfile, MAX_NAME, "%s/current_now", info->info_file);
    if ((buf = get_acpi_content(sysfile)) != NULL)
        info->present_rate = strtol(buf, NULL, 10);
    free(buf);

	/* get information from the info file */
	batt_charge_state(info);
	
	return SUCCESS;
}

/* calculate percentage of battery capacity num */
static void
calc_remain_perc(const int num){
	float lfcap;
	battery_t *info = &batteries[num];
	int perc;

	if(info->remaining_cap < 0){
		info->percentage = NOT_SUPPORTED;
		return;
	}
	else{
		lfcap = info->last_full_cap;
		if(lfcap <= 0) lfcap = 1;
		perc = (int) ((info->remaining_cap / lfcap) * 100.0);
	}
	info->percentage = perc > 100 ? 100 : perc;
}

/* calculate remaining charge time for battery num */
static void
calc_remain_chargetime(const int num){
	battery_t *info = &batteries[num];

	if(info->present_rate < 0 || info->charge_state != C_CHARGE){
		info->charge_time = 0;
		return;
	}
	info->charge_time = (int) ((((float)info->last_full_cap - (float)info->remaining_cap) / info->present_rate) * 60.0);
}

/* calculate remaining time for battery num */
static void
calc_remain_time(const int num){
	battery_t *info = &batteries[num];

	if(info->present_rate < 0 || info->charge_state != C_DISCHARGE){
		info->remaining_time = 0;
		return;
	}
	info->remaining_time = (int) (((float)info->remaining_cap / (float)info->present_rate) * 60.0);
}

/* read/refresh information about a given battery num
 * returns 0 on SUCCESS, negative values on errors */
int
read_acpi_batt(const int num){
	if(num > MAX_ITEMS) return ITEM_EXCEED;
	if (read_acpi_battstate(num) == SUCCESS) {
        read_acpi_battalarm(num, 0);
        calc_remain_perc(num);
        calc_remain_chargetime(num);
        calc_remain_time(num);
        return SUCCESS;
    }
    return -1;
}
